<?xml version="1.0"?>
<launch>
    <!-- The planning and execution components of MoveIt! configured to run -->
    <!-- using the ROS-Industrial interface. -->

    <!-- | 		PARAMETERS  		| -->
    <!-- the "sim" argument controls whether we connect to a Simulated or Real robot. -->
    <arg name="robot_sim"           default="true"/>
    <arg name="camera_sim"          default="true"/>
    <arg name="use_sdh"             default="true"/>
    <arg name="use_iiwa"            default="true"/>
    <arg name="use_interbotix"      default="false"/>
    <arg name="use_calibration"     default="true"/>
    <arg name="dream_camera"        default="faLse"/>

    <!-- hardware_interface to use : PositionJointInterface, EffortJointInterface, VelocityJointInterface. -->
    <arg name="hardware_interface"  default="PositionJointInterface"/>

    <!-- Model of the iiwa to use : iiwa7, iiwa14 -->
    <arg name="ee"                  default="sdh" />
    <arg name="robot_model"         default="iiwa7"/>
    <arg name="rviz"                default="true" />

    <!-- The is gonna have its nodes/topics under a namespace with the same name. -->
    <arg name="robot_name"          default="$(arg robot_model)"/> <!-- $(arg robot_model) -->
    <arg name="camera_name"         default="camera"/>

    <!-- camera -->
    <arg name="color_width"         default="1280"/> <!-- 1920 -->
    <arg name="color_height"        default="720"/> <!-- 1080 -->

    <!-- control -->
    <arg name="arm_operating_mode"  default="position"/> <!-- velocity-->

    <!-- TODO: simulation should also work for different rgb and depth resolution-->
    <arg unless="$(arg robot_sim)" name="depth_width" default="848"/>
    <arg unless="$(arg robot_sim)" name="depth_height" default="480"/>
    <arg if="$(arg robot_sim)" name="depth_width" default="$(arg color_width)"/>
    <arg if="$(arg robot_sim)" name="depth_height" default="$(arg color_height)"/>

    <arg name="filters"             default="disparity,spatial,temporal,pointcloud"/> <!-- disparity,spatial,temporal,decimation,hole_filling,pointcloud -->

    <!-- TODO: should depend on use_iiwa, use_interbotix...-->
    <arg if="$(arg robot_sim)" name="scene_file" default="$(find flex_gazebo_iiwa)/config/simulation.scene"/>
    <arg unless="$(arg robot_sim)" name="scene_file" default="$(find flex_gazebo_iiwa)/config/experiment.scene"/>


    <remap from="/$(arg hardware_interface)_trajectory_controller/follow_joint_trajectory"
           to="/$(arg robot_name)/$(arg hardware_interface)_trajectory_controller/follow_joint_trajectory"/>

    <group ns="$(arg robot_name)">
      <param name="dream_camera"        value="$(arg dream_camera)"/>

      <param name="use_iiwa"        value="$(arg use_iiwa)"/>
      <param name="use_sdh"         value="$(arg use_sdh)"/>
      <param name="use_interbotix"  value="$(arg use_interbotix)"/>

      <param name="camera_sim"      value="$(arg camera_sim)"/>
      <param name="robot_sim"       value="$(arg robot_sim)"/>
    </group>

    <!-- TODO: When IR is publsihed, we no longer receive depth info? -->
    <include unless="$(arg camera_sim)" file="$(find realsense2_camera)/launch/rs_rgbd.launch" ns="$(arg robot_name)"> <!-- rs_aligned_depth -->
      <arg name="camera" value="$(arg camera_name)"/>
      <arg name="enable_infra1"     value="false"/>
      <arg name="enable_infra2"     value="false"/>
      <arg name="enable_pointcloud" value="true"/>
      <arg name="align_depth"       value="true"/>
      <arg name="depth_width"       value="$(arg depth_width)"/>
      <arg name="depth_height"      value="$(arg depth_height)"/>
      <arg name="color_width"       value="$(arg color_width)"/>
      <arg name="color_height"      value="$(arg color_height)"/>
      <arg name="filters"           value="$(arg filters)"/>
    </include>


    <remap if="$(arg dream_camera)" from="/$(arg robot_name)/camera/depth_registered"
      to="/$(arg robot_name)/no_camera/depth_registered"/>
    <remap if="$(arg dream_camera)" from="/$(arg robot_name)/camera/"
      to="/$(arg robot_name)/no_camera/"/>
    <remap if="$(arg dream_camera)" from="/$(arg robot_name)/camera/color/"
      to="/$(arg robot_name)/no_camera/color"/>


  <include if="$(eval arg('robot_sim') or arg('camera_sim'))" file="$(find flex_gazebo_iiwa)/launch/gazebo.launch">
      <arg name="hardware_interface"  value="$(arg hardware_interface)"/>
      <arg name="robot_name"          value="$(arg robot_name)" />
      <arg name="camera_name"         value="$(arg camera_name)"/>

      <arg name="use_sdh"             value="$(arg use_sdh)"/>
      <arg name="robot_model"         value="$(arg robot_model)"/>

      <arg name="camera_sim"          value="$(arg camera_sim)"/>
      <arg name="robot_sim"           value="$(arg robot_sim)"/>

      <arg name="depth_width"         value="$(arg depth_width)"/>
      <arg name="depth_height"        value="$(arg depth_height)"/>
      <arg name="color_width"         value="$(arg color_width)"/>
      <arg name="color_height"        value="$(arg color_height)"/>
  </include>

    <!-- Load move_group -->
    <group if="$(arg use_sdh)" ns="$(arg robot_name)">
      <!-- iiwa with SDH-->
      <include file="$(find flex_sdh_moveit)/launch/move_group.launch">
        <arg name="publish_monitored_planning_scene" value="true" />
        <arg name="hardware_interface" value="$(arg hardware_interface)"/>
	      <arg name="robot_name" value="$(arg robot_name)"/>
        <arg name="robot_model" value="$(arg robot_model)" />

        <arg name="camera_name" value="$(arg camera_name)"/>
        <arg name="camera_sim" value="$(arg camera_sim)"/>

        <arg name="use_sdh" value="$(arg use_sdh)"/>
        <arg name="use_iiwa" default="$(arg use_iiwa)"/>
        <arg name="use_interbotix" default="$(arg use_interbotix)"/>
      </include>

      <include if="$(arg rviz)" file="$(find flex_sdh_moveit)/launch/moveit_rviz.launch">
        <arg name="config" value="true"/>
      </include>
    </group>
    <group unless="$(arg use_sdh)" ns="$(arg robot_name)">
      <!-- iiwa without SDH-->
      <include file="$(find flex_gripper_moveit)/launch/move_group.launch">
        <arg name="publish_monitored_planning_scene" value="true" />
        <arg name="hardware_interface" value="$(arg hardware_interface)"/>
        <arg name="robot_name" value="$(arg robot_name)"/>
        <arg name="model" value="$(arg robot_model)" />
      </include>

      <include if="$(arg rviz)" file="$(find flex_gripper_moveit)/launch/moveit_rviz.launch">
        <arg name="config" value="true"/>
      </include>
    </group>

    <arg name="use_actual" if="$(arg robot_sim)" value="false"/>
    <arg name="use_actual" unless="$(arg robot_sim)" value="true"/>


    <!-- Working with the real robot.  -->
    <group if="$(arg use_actual)" ns="$(arg robot_name)">
        <!-- Load controllers accordingly to parameters -->
        <include file="$(find iiwa_control)/launch/iiwa_control.launch">
            <arg name="hardware_interface" value="$(arg hardware_interface)" />
            <arg name="controllers" value="joint_state_controller $(arg hardware_interface)_trajectory_controller" />
            <arg name="robot_name" value="$(arg robot_name)" />
            <arg name="model" value="$(arg robot_model)" />
        </include>

        <!-- Robot interface -->
        <include file="$(find iiwa_hw)/launch/iiwa_hw.launch" >
            <arg name="hardware_interface" value="$(arg hardware_interface)" />
        </include>
    </group>

    <node name = "moveit_publish_scene_from_text"
      pkg= "moveit_ros_planning"
      type = "moveit_publish_scene_from_text"
      args= "$(arg scene_file)"
      ns="$(arg robot_name)"/>


    <include if="$(arg use_calibration)" file="$(find easy_handeye)/launch/publish.launch">
      <arg name="eye_on_hand"                       value="false"/>
      <arg name="namespace_prefix"                  value="calibration"/>

    </include>

</launch>
